DIMO-Network/dimo-python-sdk
# Initial SDK Review

This is an initial, large-ish PR for the DIMO Python SDK. Currently, it does not yet include support for Streamr, nor tests/detailed error messages, but those are planned for a later date. It is also not yet fully configured for package upload to PyPi (currently working thru testing this on TestPyPi), but those changes will be made once the initial SDK has been reviewed and approved by the DIMO team. 

**How To Test (for now, pre-PyPi):**
1. Clone the repo locally and switch to the `dev-barrettk` branch & pull.
2. In your terminal, `cd` into the `dimo` folder
3. Create a new file called `test.py` inside the `dimo` folder and import the following (as needed) and initialize dimo
```python
from dimo import DIMO
import asyncio
from dotenv import load_dotenv # Only needed if planning to add a private key into your .env for testing
import os # Only needed if planning to add a private key into your .env for testing 

dimo = DIMO("Production")
```
4. Inside `test.py`, create an async method following this format (see considerations note below):    
```python
async def main():

  #Your test code here

if __name__ == "__main__":
        asyncio.run(main())
```
5. Use the README to follow along/test different endpoints. You can run your file in the terminal by running `python test.py`. If you're using a `.env` for your private key, and you're testing the `get_token()` method, your code might look something like this:

```python
async def main():

    authHeader = await dimo.auth.get_token(
                client_id='<client_id>',
                domain='<domain>',
                private_key=os.getenv("test_private_key")
    )
    print("Full Auth Header: ", authHeader)
    access_token = authHeader["access_token"]
    print("My access_token: ", access_token)

if __name__ == "__main__":
        asyncio.run(main())
```

**Considerations/Notes:**
* Currently, the SDK uses [asyncio](https://docs.python.org/3/library/asyncio.html) to make asynchronous requests for all endpoints, in an attempt to generally follow the same patterns as the NodeJS SDK. This especially makes sense for methods such as `get_token()`, but may not be required for other endpoints. Would love to get thoughts/feedback on this, as I’d like to make the SDK as accessible/easy to use for devs as possible. 
* The SDK currently uses the same template graphQL queries for the Identity & Telemetry classes as the NodeJS SDK, but more can be added. 
* README is mostly complete with some placeholders on items that are “Coming Soon”. Note that the pip installation is also placeholder and is not yet functional.
